tasks/src/lib.rs

```diff
@@ -0,0 +1,323 @@
+    ///
+    /// Note: The [`Scheduler`] will tag the [`CommandBody`] creating a complete [`Command`].
+    fn command_body(&self) -> CommandBody<'static>;
+
+    /// Process data response.
+    fn process_data(&mut self, data: Data<'static>) -> Option<Data<'static>> {
+        // Default: Don't process server data
+        Some(data)
+    }
+
+    /// Process an untagged response.
+    fn process_untagged(
+        &mut self,
+        status_body: StatusBody<'static>,
+    ) -> Option<StatusBody<'static>> {
+        // Default: Don't process untagged status
+        Some(status_body)
+    }
+
+    /// Process bye response.
+    fn process_bye(&mut self, bye: Bye<'static>) -> Option<Bye<'static>> {
+        // Default: Don't process bye
+        Some(bye)
+    }
+
+    /// Process command completion result response.
+    ///
+    /// The [`Scheduler`] already chooses the corresponding response by tag.
+    fn process_tagged(self, status_body: StatusBody<'static>) -> Self::Output;
+}
+
+/// Scheduler.
+pub struct Scheduler {
+    flow: ClientFlow,
+    waiting_tasks: HashMap<ClientFlowCommandHandle, Box<dyn TaskAny>>,
+    active_tasks: HashMap<ClientFlowCommandHandle, (Tag<'static>, Box<dyn TaskAny>)>,
+    tag_generator: TagGenerator,
+}
+
+impl Scheduler {
+    pub fn new(flow: ClientFlow) -> Self {
+        Self {
+            flow,
+            waiting_tasks: Default::default(),
+            active_tasks: Default::default(),
+            tag_generator: TagGenerator::new(),
+        }
+    }
+
+    pub fn enqueue<T>(&mut self, task: T) -> TaskHandle<T>
+    where
+        T: Task,
+    {
+        let cmd = {
+            let tag = self.tag_generator.generate();
+            let body = task.command_body();
+
+            Command { tag, body }
+        };
+
+        let handle = self.flow.enqueue_command(cmd);
+
+        let replaced = self.waiting_tasks.insert(handle, Box::new(task));
+        assert!(replaced.is_none());
+
+        TaskHandle::new(handle)
+    }
+
+    pub async fn progress(&mut self) -> SchedulerEvent {
+        loop {
+            // TODO: How to handle error?
+            let event = self.flow.progress().await.unwrap();
+
+            match event {
+                ClientFlowEvent::CommandSent { tag, handle } => {
+                    // Safety: `unwrap` can't fail here.
+                    let entry = self.waiting_tasks.remove(&handle).unwrap();
+                    self.active_tasks.insert(handle, (tag, entry));
+                }
+                ClientFlowEvent::CommandRejected { tag, status, .. } => {
+                    let handle = {
+                        let (handle, _) = self
+                            .active_tasks
+                            .iter()
+                            .find(|(_, (tag_, _))| tag == *tag_)
+                            .unwrap();
+
+                        *handle
+                    };
+
+                    let body = match StatusV2::from(status) {
```

Comment:
I commited `StatusV2` to imap-codec. `Status` and `StatusV2` are isomorph -- tight types, yay! -- but V2 is much more pleasent to work with. Not sure how to continue with this...